cpp/src/arrow/compute/kernels/aggregate.h

```diff
@@ -0,0 +1,111 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+
+#include "arrow/compute/kernel.h"
+
+namespace arrow {
+
+class Array;
+class Status;
+
+namespace compute {
+
+class FunctionContext;
+struct Datum;
+
+/// AggregateFunction is an interface for Aggregates
+///
+/// An aggregates transforms an array into single result called a state via the
+/// Consume method.. State supports the merge operation via the Merge method.
+/// State can be sealed into a final result via the Finalize method.
+//
+/// State ownership is handled by callers, thus the interface exposes 3 methods
+/// for the caller to manage memory:
+/// - Size
+/// - New (placement new constructor invocation)
+/// - Delete (state desctructor)
+///
+/// Design borrowed from ClickHouse aggregate functions.
+class AggregateFunction {
+ public:
+  /// \brief Consume an array into a state.
+  virtual Status Consume(const Array& input, void* state) const = 0;
+
+  /// \brief Merge states.
+  virtual Status Merge(const void* src, void* dst) const = 0;
+
+  /// \brief Convert state into a final result.
+  virtual Status Finalize(const void* src, Datum* output) const = 0;
+
+  virtual ~AggregateFunction() {}
+
+  /// State management methods.
+  virtual int64_t Size() const = 0;
+  virtual void New(void* ptr) const = 0;
+  virtual void Delete(void* ptr) const = 0;
```

Comment:
Because you do not wanted shared-ptr boxed primitives for group-by results, you write into a Buffer (which can then be converted to a proper typed Array). This is critical for cases where the group expression does not reduce the data size.

The second issue is that some operators requires extra allocation for each results (`think HyperLogLog(precision)`, or `Distribution(max_vals))`. This poses 2 sub-problems, the size of the state may require extra allocation (and thus reclamation via destructor), and the size might be a function of a non-static argument.

The following query would be untenable with `shared_ptr<State>`:

```SQL
SELECT
  campaign_id,
  user_id,
  SUM(impressions),
  SUM(clicks),
  HYPERLOGLOG(sites, bins = 64, precision_bits=8), -- size is 64 bytes
  DISTRIBUTION(sites, top_bins = 50) -- size is struct<String, size_t>[50]
GROUP BY   -- Very high cardinality groups
  campaign_id,
  user_id; 
```

Reply:
I still don't understand the need for the elaborate indirection. I'm sorry to be a bit thick, let's address it once we implement some more complex aggregation functions